### PR TITLE
Handle group messages

### LINKED ISSUE #13: 支援 LINE chatbot 群組聊天
LINE Bot 再被加進去群組之後，把在群組裡面的訊息，拿去對前 100 則查證的訊息做 in-memory 的查詢（#14）。

Related discussion:
20170318 - https://beta.hackfoldr.org/cofacts/https%253A%252F%252Fhackmd.io%252Fs%252FrJ0nBnYsx

### PR DESCRIPTION
**Pull request by nonumpa**
Fixes #13

### Not ready
- (done)~~Change validCategories from categoryName to categoryId~~
- (done)~~Should filter out the category which negative vote > positive vote~~
- Self-introduction wording
- (done)~~Should disable job queue rate limit and add `concurrency` to env variable~~
- Maybe add a leave group command, it seems people can't kick others in the room

### Flow
webhook/index (get group message event) ->  GroupHandler(jobQueue) -> processGroupEvent -> groupMessage

### Response rule
- **Chat-bot self-introduction** : Input `Hi cofacts` or `Hi confacts` .
- **Ignore message** : Input length less than 10 won't reply and post to cofacts-api.
- **Reply** :
    - The article found should be in `validCategories` and be **Identical enough**.
        - `validCategories`: `疾病、醫藥🆕`, `COVID-19 疫情🆕`, `科技、資安、隱私`, `保健秘訣、食品安全`, `免費訊息詐騙`
    - The reply of the article should:
        1. Candidate's `positiveFeedbackCount` > candidate's `negativeFeedbackCount.`
        2. Candidate's `positiveFeedbackCount` > non-rumors' `positiveFeedbackCount`. 
        3. If replyCount > 2, check if rumor count >> non-rumor count (2:1).

### Job queue
- `jobQueue` will process events which `replytoken` are not expired.
- If the event which `jobQueue` is going to process is expired, it will be skipped and added to `expiredJobQueue`.
- `expiredJobQueue` **start** only when there's no jobs in `jobQueue`(drained).
- `expiredJobQueue` **pause** when there's jobs adding to `jobQueue`.

#### Add two user scope custom dimension
1. `Message Source`: `user` | `group` | `room` will set every time sends the event.
2. `NumberOfGroupMembers`:  It will only set when bot join group.

#### Article event in group
Category / Action / Label
- If user trigger chat-bot to introduce self:
    - `UserInput` /  `Intro` / 
- If we found a articles in database that matches the message:
    - `UserInput` / `ArticleSearch` / `ArticleFound`
    - `Article` / `Search` / `<article id>` for each article found
- We don't send event If nothing found in database

#### To calculate total group the bot joined, we set ga event value 1 as join, -1 as leave
 Category / Action / Value 

- Join Event
    - 'Group' /  'Join' / 1
- Leave Event
    - 'Group' /  'Leave' / -1

Snapshot shows where to create custom dimension in ga.

截圖 2021-01-26 下午7 24 26

### CODE REVIEW

**Review by MrOrz — COMMENTED** [review]
I'm still reviewing, just send out 2 comments here

File: src/webhook/handlers/__tests__/groupHandler.test.js [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+  gh.addJob(param, { jobId });
+
+  jobQueue.on('failed', async (job, e) => {
+    // console.log('jobQueue.failed');
+    expect(job.id).toBe(jobId);
+    expect(e instanceof TimeoutError).toBe(true);
+    expect(e).toMatchInlineSnapshot(`[Error: Event expired]`);
+  });
+
+  // in real case this will not happen
```
Comment (MrOrz) [inline src/webhook/handlers/__tests__/groupHandler.test.js:230]: It seems that it should happen in real case as well.

When a job is added to `expiredJobQueue` in `processJob`, it must have an expired timestamp.
By the time the job is executed by `expiredJobQueue`, `processGroupEvent` will be invoked, and `isEventExpired()` inside will still throw a `TimeoutError`.

File: src/webhook/handlers/__tests__/groupHandler.test.js [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+    // console.log('processGroupEvent failed with timeout');
+    return Promise.reject(
+      new TimeoutError('mock processGroupEvent timeout error')
+    );
+  });
+  const gh = new GroupHandler(jobQueue, expiredJobQueue, 1);
+  gh.addJob(param, { jobId: 'testExpiredId1' });
+  gh.addJob(param, { jobId: 'testExpiredId2' });
+  gh.addJob(param, { jobId: 'testExpiredId3' });
+  gh.addJob(param, { jobId: 'testExpiredId4' });
```
Comment (MrOrz) [inline src/webhook/handlers/__tests__/groupHandler.test.js:298]: I am confused by the 4 jobs and the changing `processGroupEvent` mock. May I know what would happen to the 4 jobs here?

**Review by MrOrz — CHANGES_REQUESTED** [review]
Thanks for the implementation!

I have added some comments and some hints on GA. Happy Lunar New Year!

File: src/webhook/handlers/__tests__/groupHandler.test.js (hunk 2 above)
Comment (nonumpa) [reply src/webhook/handlers/__tests__/groupHandler.test.js:298]: 有點難描述，我用中文好了 哈哈

這裡要測試 expireQueue 執行時，如果有新的 event 進來，expireQueue 要被暫停，並且執行完 jobQueue 後再繼續，利用監聽 expiredJobQueue 的 active event 加上 `concurrency` 設成 `1`，我們可以確保在 expiredJobQueue 執行期間，插入一個沒過期的 job。
1. 因為這裡把 `GroupHandler` 的 `concurrency` 設成 `1`，所以四個 expiredJob 會在 jobQueue 的 process function [`processJob`](https://github.com/cofacts/rumors-line-bot/pull/238/files/a04a06cccbb584af1d0e3e946e6e8c1b4ff012db#diff-eea0db6bb6c18528afc0978099a6d13f4529cf3f945e332e7af1749204c39d92R38) 一個一個(不保證順序)處理並加進 expiredJobQueue，這裡並不會執行到 `processJob` 裡面的 `processGroupEvent`。
2. 當所有(四個) jobQueue 的 job 被執行完之後(`onDrained`)，expiredJobQueue 會開始(resume)執行。
3. 當第一個 expired job 開始執行時，會同時送出一個 active event。
4. 在收到 active event 的時候加入一個沒過期的 job，並且為了讓它成功，把 `processGroupEvent` 改成 resolve，在complete 改回來。
其實這邊我期望是可以 mockImplementationOnce(expired x 4 + success x 1)，但是 `processGroupEvent` 被執行的順序可能是 
    - testExpiredId1 -> (expiredJobQueue active event) -> successJobId -> testExpiredId2 -> ...
    - testExpiredId1 -> (expiredJobQueue active event) testExpiredId2 （幾乎同時）-> successJobId ->  ...
  
    因為我們不能保證 active event 收到的時機，不知道 success mock 要放在第幾個順位才能正確讓 successJobId 是 resolve 的，所以就在收到第一個 expiredJobQueue (job) 的 active event 時把它改成 resolve，在它完成時再改回 reject，這樣就會造成一個問題是：把 `processGroupEvent` 改成 resolve 時剛好 testExpiredId2 開始了 expiredJobQueue 還沒 pause，所以 testExpiredId2 的結果就是 complete 的(理論上 expiredJobQueue 的 job 要在 `processGroupEvent` 裡面全被 reject)
這裡只 expect jobQueue 的 success/fail count，因為 expiredJobQueue 的結果可能每次都不一樣

Comment (MrOrz) [reply src/webhook/handlers/__tests__/groupHandler.test.js:298]: 感謝感謝
我在想如果「這裡要測試 expireQueue 執行時，如果有新的 event 進來，expireQueue 要被暫停，並且執行完 jobQueue 後再繼續」

若把下面這兩塊拆成兩個分開的 test case，會不會比較簡單
1.  `expireQueue` 執行時，如果有新的 event 進來，`expireQueue` 要被暫停 --> 先往 `expireQueue` 塞東西，再 `gh.addJob`，然後 expect `expireQueue` 的 status
2. 執行完 `jobQueue` 後再繼續 --> 先準備一個暫停的 expireQueue，然後 `gh.addJob`，然後在該 job 完成之後 expect expireQueue 的 status 是否不是 pause

File: src/webhook/handlers/__tests__/groupHandler.test.js (hunk 1 above)
Comment (nonumpa) [reply src/webhook/handlers/__tests__/groupHandler.test.js:226]: Sorry, the comment may come from copy and paste and I forget to remove it.

**Review by MrOrz — APPROVED** [review]
Seems that most issues are resolved. Things left are
1. documentation regarding the custom dimensions; and if we should use custom metric instead of custom dimension for user count in chat groups
2. excessive comment
3. simplify test case

I think we can discuss on 1 because setting custom metric / dimension cannot be undone in the future. Documentations, comments & tests can wait.

**Comment by MrOrz** [comment]
We will merge first and just setup 1st custom dimension. Leave 2nd dimension / metric for the future.

**Pull request merged**